fineract-provider/build.gradle

```diff
@@ -364,7 +363,7 @@ task integrationTest(type:Test){
         tomcatRunWar.daemon = true
         tomcatRunWar.execute()
```

Comment:
With the removal of TaskInternal.execute() method in Gradle 5.x , this task (integrationTest) would fail

Reply:
Also all flyway tasks fail since they are using `TaskInternal.execute()`

Reply:
I tinkered a bit using the specified `gradle 5.2.1` and discovered that the specified flyway plugin `3.0` is not compatible with the gradle version due to the above mentioned error. I discovered that in flyway `3.1` they added a new feature where one can override `FlywayTask` so implementing the flyway tasks using the below format would get rid of the `TaskInternal.execute()` error:
```

task migrateTenantListDB(type: org.flywaydb.gradle.task.FlywayMigrateTask) {
    description="Migrates a Tenant List DB. Optionally can pass dbName. Defaults to 'mifosplatform-tenants' (Example: -PdbName=someDBname)"
    def filePath = "filesystem:$projectDir" + '/src/main/resources/sql/migrations/list_db'
    def tenantsDbName = 'mifosplatform-tenants';
    if (rootProject.hasProperty("dbName")) {
        tenantsDbName = rootProject.getProperty("dbName")
    }
    extension = new org.flywaydb.gradle.FlywayExtension()
    extension.url= "jdbc:mysql:thin://localhost:3306/$tenantsDbName"
    extension.driver = "org.drizzle.jdbc.DrizzleDriver"
    extension.user = 'root'
    extension.password = '.....'
    extension.locations= [filePath]
}
```

However, in gradle `5.0` the property `sourceSet.output.classesDir` was removed in favour of `sourceSet.output.classesDirs`. You'll notice that in the internal implementation of Flyway 3.1 [here](https://github.com/flyway/flyway/blob/f7fa24196789e2cbcc5ffffb409b5eca2c629846/flyway-gradle-plugin/src/main/groovy/org/flywaydb/gradle/task/AbstractFlywayTask.groovy) the propety `classesDir` is still being used:
```
def runTask() {
        if (isJavaProject()) {
            def classLoader = Thread.currentThread().getContextClassLoader()
            project.sourceSets.each {
                def classesUrl = it.output.classesDir.toURI().toURL() <----------
                logger.debug("Adding directory to Classpath: " + classesUrl)
                classLoader.addURL(classesUrl)
        ...............
}
```
Flyway `5.0.0` was corrected to support both `classesDir` and `classesDirs`, however, it doesn't seem to be compatible with `drizzle-jdbc`; even the latest version `1.4`. I tried the default `mysql-connector` but it always threw a bunch of errors during migration. Some are show below:
```
DB: Unknown table 'acc_gl_account' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'acc_gl_closure' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'acc_gl_journal_entry' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'acc_product_mapping' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'c_configuration' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'm_appuser' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'm_appuser_role' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'm_calendar' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'm_calendar_instance' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'm_charge' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'm_client' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'm_client_identifier' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'm_code' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'm_code_value' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'm_currency' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'm_deposit_account' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'm_deposit_account_transaction' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'm_document' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'm_fund' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'm_group' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'm_group_level' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'm_group_client' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'm_guarantor' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'm_loan' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'm_loan_charge' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'm_loan_arrears_aging' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'm_loan_collateral' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'm_loan_officer_assignment_history' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'm_loan_repayment_schedule' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'm_loan_transaction' (SQL State: 42S02 - Error Code: 1051)
DB: No data - zero rows fetched, selected, or processed (SQL State: 02000 - Error Code: 1329)
DB: No data - zero rows fetched, selected, or processed (SQL State: 02000 - Error Code: 1329)
DB: No data - zero rows fetched, selected, or processed (SQL State: 02000 - Error Code: 1329)
DB: No data - zero rows fetched, selected, or processed (SQL State: 02000 - Error Code: 1329)
DB: Unknown table 'm_template' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'm_templatemappers' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'm_template_m_templatemappers' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'sms_messages_outbound' (SQL State: 42S02 - Error Code: 1051)
DB: Data truncated for column 'last_time_password_updated' at row 1 (SQL State: 01000 - Error Code: 1265)
DB: You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '--Journal entries also refer to Client transactions

```
 To remove those errors changes will have to be made the `.sql` files.

Reply:
@michaelbukachi following this and it is a quite interesting discovery for me. I have pulled this in and switched to a new branch. So i can work on this . Have been following all along and i see some changes have to be made to the sql file. 

What changes are to be made. I am a newbie and this depends on an issue i am trying to solve for integrating spotBugs - here https://issues.apache.org/jira/browse/FINERACT-702 which depends on a latest version of gradle.

I would love to help in removing unwanted stuff or update the sql file accordingly thanks

Reply:
These are the changes which need to made:

- Updating of the sql files to support latest mysql/mariadb versions. Note. This can be quite tedious due to the fact there are around 360+ `.sql` files to go through.
- Once that is done, you'll need to change the drizzle driver to something else. You'll have to choose between mysql and mariadb jdbc drivers. Note. Drizzle is used all over the place in different files so you might have to do a search
- Next, you'll have to update the flyway plugin
- Finally, you can update gradle to the latest version
- Remember to run integration tests in the end to make sure everything is good.

Reply:
Thanks alot. I am on it. will ping you if i have an error

Reply:
> pulled this in and switched to a new branch. So i can work on this

@muarachmann if it's possible to raise small incremental PRs, e.g. first doing some SQL migrations, if they are backwards compatible, it makes it easier for other people to review smaller changes and merge them gradually.

> choose between mysql and mariadb jdbc drivers

IMHO we should use the [MariaDB JDBC Driver](https://mariadb.com/kb/en/library/about-mariadb-connector-j/), not the (Oracle..) MySQL driver. I know less about Drizzle, but thought that it was abandoned; note drizzle.org is dead & last org.drizzle.jdbc:drizzle-jdbc released in Aug, 2016, so I would ignore that. Again as I said above, it may be best to do this in separate gradual incremental Pull Requests.

Reply:
@vorburger Is the MariaDB driver license compatible with the Apache License? The ASF prohibits the use of non compliant (category X) libraries on Apache projects. Maybe there is the work around for this? @myrle-krantz , @vishwasbabu

Reply:
@awasum seems MariaDB driver is under the LGPL license which [seems](https://www-us.apache.org/legal/resolved.html#category-x) not to be compatible with ASF

Reply:
@michaelbukachi I spoke with @muarachmann over the phone and he's trying to update .sql files to be compliant with mysql 8. I'm not really clear as to how upgrading gradle to 5.2.1 will require a mysql upgrade. I understand that the drizzle plugin is not compatible, but shouldn't we just be replacing it with a compatible plugin that still maintains the current mysql syntax?

> Updating of the sql files to support latest mysql/mariadb versions. Note. This can be quite tedious due to the fact there are around 360+ .sql files to go through.

I feel the above should have been in a separate issue IMO.

Reply:
@ivange94 It's because the flyway plugin is affected. There current plugin isn't compatible with gradle 5. Upgrading the plugin would require upgrading the driver.

Reply:
> @michaelbukachi I spoke with @muarachmann over the phone and he's trying to update .sql files to be compliant with mysql 8. I'm not really clear as to how upgrading gradle to 5.2.1 will require a mysql upgrade. I understand that the drizzle plugin is not compatible, but shouldn't we just be replacing it with a compatible plugin that still maintains the current mysql syntax?
> 
> > Updating of the sql files to support latest mysql/mariadb versions. Note. This can be quite tedious due to the fact there are around 360+ .sql files to go through.
> 
> I feel the above should have been in a separate issue IMO.

I think he said he didn't find a compatible drizzle-jdbc plugin. But @muarachmann can still try to do some research on getting a compatible drizzle plugin

Reply:
Resolving this old and now outdated conversation, because it's no longer relevant for this PR.